net/decentralized_dns/constants.h

```diff
@@ -6,21 +6,72 @@
 #ifndef BRAVE_NET_DECENTRALIZED_DNS_CONSTANTS_H_
 #define BRAVE_NET_DECENTRALIZED_DNS_CONSTANTS_H_
 
+#include <string_view>
+
+#include "base/containers/fixed_flat_set.h"
+
 namespace decentralized_dns {
 
-inline constexpr const char* kUnstoppableDomains[] = {
-    ".crypto",     ".x",          ".nft",     ".dao",         ".wallet",
-    ".blockchain", ".bitcoin",    ".zil",     ".altimist",    ".anime",
-    ".klever",     ".manga",      ".polygon", ".unstoppable", ".pudgy",
-    ".tball",      ".stepn",      ".secret",  ".raiin",       ".pog",
-    ".clay",       ".metropolis", ".witg",    ".ubu",         ".kryptic",
-    ".farms",      ".dfz",        ".kresus",  ".binanceus",   ".austin",
-    ".bitget",     ".wrkx"};
+// A struct to be used to guide the lookup of domain suffixes in the fixed flat
+// set below. Lookups using this wrapper will match based on the suffix after
+// the last '.' in the domain. Lookups for suffix can be done like:
+//
+// kUnstoppableDomains.contains(DomainSuffixLookup("https://foo.crypto"))
+//
+struct DomainSuffixLookup {
+  constexpr explicit DomainSuffixLookup(std::string_view domain)
+      : domain(ExtractSuffix(domain)) {}
+
+  // Used to initialise the view from the last '.' in the domain, if any exist.
+  // Otherwise, passes the whole string for comparison.
+  static std::string_view ExtractSuffix(std::string_view domain) {
+    size_t pos = domain.find_last_of('.');
+    return pos != std::string_view::npos ? domain.substr(pos) : domain;
+  }
+
+  std::string_view domain;
+};
+
+namespace internal {
+
+// A custom comparator with strict weak ordering, to allow the lookup of a
+// domain with `DomainSuffixLookup`.
+struct SuffixComparator {
+  using is_transparent = void;
+
+  constexpr bool operator()(const std::string_view& lhs,
+                            const std::string_view& rhs) const {
+    return lhs < rhs;
+  }
+
+  constexpr bool operator()(const std::string_view& lhs,
+                            DomainSuffixLookup rhs) const {
+    return (*this)(lhs, rhs.domain);
+  }
+
+  constexpr bool operator()(DomainSuffixLookup lhs,
+                            const std::string_view& rhs) const {
+    return (*this)(lhs.domain, rhs);
+  }
+};
```

Comment:
tbh I'd put all this machinery in `.cc` (including `kUnstoppableDomains` declaration), and export a single function like `std::optional<std::string_view> GetUnstoppableDomainSuffix(std::string_view);`.

Reply:
I can do that but it will run into problems with this elsewhere:

```cxx
base::ASCIIToUTF16(base::JoinString(kUnstoppableDomains, ", ")),
```

Reply:
Fixed. I had to create `GetUnstoppableDomainSuffixFullList`, which in my opinion is worse for readability of the code in general, but eh.

Reply:
> Fixed. I had to create `GetUnstoppableDomainSuffixFullList`, which in my opinion is worse for readability of the code in general, but eh.

that seems appropriate here. Having a big array of domains be inlined into multiple functions looks worse imo.

I'm actually curious if they are deduplicated or not by the linker/compiler. I expect them to be duplicated.

Reply:
That's a good question. As I understand it, once you tag a global as `inline` that guarantees a single address.